File: 프로그래머스/2020 KAKAO BLIND RECRUITMENT/문자열 압축.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int solution(string s) {
   int answer = s.length();
   for(int i = 1; i <= s.length()/2; ++i){
     int pos = 0;
     int len = s.length();

     while(true){
       string unit = s.substr(pos,i);
       pos += i;
       if(pos >= s.length()) break;

       int cnt = 0;
       while(true){
           // 같은 문자열인 경우
         if(unit.compare(s.substr(pos,i)) == 0){
           ++cnt;
           pos += i;
         }
         else break;
       }
       if(cnt > 0){
         len -= i * cnt;
         if(cnt < 9) len += 1;
         else if(cnt < 99) len += 2;
         else if(cnt < 999) len += 3;
         else len += 4;
       }
     }
     answer = min(answer,len);
   }
   return answer;
}

File: 프로그래머스/2020 KAKAO BLIND RECRUITMENT/문자열 압축.js
function solution(s) {
    let answer = 1000;
    if(s.length == 1){
        return 1
    }
    for(let i = 1; i < parseInt(s.length / 2) +1 ; i++){
        let result = '';
        let before = '';
        let cnt = 1;
        for(let j = 0; j < s.length; j += i){
            if(before === s.slice(j,j+i)){
                cnt += 1;
            }else if(j != 0){
                if(cnt > 1){
                    result += cnt;
                }
                result += before;
                before = s.slice(j,j+i);
                cnt = 1;
            }
            else{
                before = s.slice(j,j+i);
            }
        }
        if(cnt > 1){
            result += cnt;
        }
        result += before;
        if (answer > result.length){
            answer = result.length;
        }
    }
    return answer;
}